lib/MongoDB/Database.pm

```diff
@@ -409,27 +412,68 @@ on database commands: L<http://dochub.mongodb.org/core/commands>.
 =cut
 
 sub run_command {
-    my ( $self, $command, $read_pref ) = @_;
+    my ( $self, $command, $read_pref, $options ) = @_;
     MongoDB::UsageError->throw("command was not an ordered document")
        if ! is_OrderedDoc($command);
 
     $read_pref = MongoDB::ReadPreference->new(
         ref($read_pref) ? $read_pref : ( mode => $read_pref ) )
       if $read_pref && ref($read_pref) ne 'MongoDB::ReadPreference';
 
+    $options ||= {};
+
     my $op = MongoDB::Op::_Command->_new(
+        client      => $self->_client,
         db_name     => $self->name,
         query       => $command,
         query_flags => {},
         bson_codec  => $self->bson_codec,
         read_preference => $read_pref,
+        ( defined $options->{session} ? ( session => $options->{session} ) : () ),
     );
 
     my $obj = $self->_client->send_read_op($op);
 
     return $obj->output;
 }
 
+sub _aggregate {
+    MongoDB::UsageError->throw("pipeline argument must be an array reference")
+      unless ref( $_[1] ) eq 'ARRAY';
+
+    my ( $self, $pipeline, $options ) = @_;
+    $options ||= {};
+
+    my $session = delete $options->{session};
+
+    # boolify some options
+    for my $k (qw/allowDiskUse explain/) {
+        $options->{$k} = ( $options->{$k} ? true : false ) if exists $options->{$k};
+    }
+
+    # possibly fallback to default maxTimeMS
+    if ( ! exists $options->{maxTimeMS} && $self->max_time_ms ) {
+        $options->{maxTimeMS} = $self->max_time_ms;
+    }
+
+    # read preferences are ignored if the last stage is $out
+    my ($last_op) = keys %{ $pipeline->[-1] };
+
+    my $op = MongoDB::Op::_Aggregate->_new(
+        pipeline     => $pipeline,
+        options      => $options,
+        read_concern => $self->read_concern,
+        has_out      => $last_op eq '$out',
+        client       => $self->_client,
+        bson_codec   => $self->bson_codec,
+        db_name      => $self->name,
+        coll_name    => 1, # Magic not-an-actual-collection number
```

Comment:
If we're going to do this, we should update Op::_Aggregate to error if coll_name is "1" and wire version isn't high enough.

Reply:
Ok, have added that in the execute in Op::_Aggregate